bench/pixcmp.1
'\" t
.\"     Title: PIXCMP
.\"    Author:
.\" Generator: DocBook XSL-NS Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12/04/2012
.\"    Manual: BRL-CAD
.\"    Source: BRL-CAD
.\"  Language: English
.\"
.TH "PIXCMP" "1" "12/04/2012" "BRL\-CAD" "BRL\-CAD"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pixcmp \- compare two pix image files pixel by pixel
.SH "SYNOPSIS"
.HP \w'\fBpixcmp\fR\ 'u
\fBpixcmp\fR [\fIOPTIONS\fR] \fIFILE1\fR [\fIFILE2\fR\ [\fISKIP1\fR\ [\fISKIP2\fR]]]
.SH "DESCRIPTION"
.PP
\fBpixcmp\fR
is a program to compare two BRL\-CAD pix image files pixel by pixel\&. The following
\fIOPTIONS\fR
are available:
.PP
\fB\-l\fR
.RS 4
Output pixel numbers, counting from 1, and values for all pixels that differ\&.
.RE
.PP
\fB\-b\fR
.RS 4
Use bytes instead of pixels for both processing and output\&. With this option, the
\fB\-i\fR
SKIP values should be specified as bytes instead of pixels\&.
.RE
.PP
\fB\-i\fR \fISKIP\fR
.RS 4
Skip the first SKIP pixels of input (for FILE1 and FILE2)\&.
.RE
.PP
\fB\-i\fR \fISKIP1\fR:\fISKIP2\fR
.RS 4
Skip the first SKIP1 pixels of FILE1 and the first SKIP2 pixels of FILE2\&.
.RE
.PP
\fB\-s\fR
.RS 4
Silent output\&. Only return an exit status\&.
.RE
.PP
SKIP1 and SKIP2 are the number of pixels to skip in each file\&.
.PP
If FILE is `\fB\-\fR` or missing,
\fBpixcmp\fR
reads from the standard input\&.
.SH "RETURN VALUES"
.PP
The
\fBpixcmp\fR
utility returns
0
if there are no differences,
1
if there are only off\-by\-one differences,
2
if there are off\-by\-many errors,
126
if there are file processing problems, and
127
if there are argument processing or usage errors\&.
.SH "EXAMPLES"
.PP
\fBpixcmp \-l file1 file2\fR
.RS 4
The two files are compared pixel by pixel with each difference printed instead of just the summary\&.
.RE
.PP
\fBpixdiff file1\&.pix file2\&.pix | pixcmp \- file3\&.pix\fR
.RS 4
The
\fIpixdiff\fR
tool compares the pixels in file1\&.pix with those in file2\&.pix and then outputs a resulting `diff` image which is then processed by
\fBpixcmp\fR
as input and compared against file3\&.pix, reporting on the differences\&.
.RE
.SH "SEE ALSO"
.PP
\fBbrlcad\fR(1),
\fBpixdiff\fR(1),
\fBpix\fR(5),
\fBbw\fR(5)
.SH "AUTHOR"
.PP
Christopher Sean Morrison
Charles M\&. Kennedy
Michael John Muuss
.SH "COPYRIGHT"
.PP
This software is Copyright (c) 2007\-2012 by the United States Government as represented by U\&.S\&. Army Research Laboratory\&.
.SH "BUG REPORTS"
.PP
Reports of bugs or problems should be submitted via electronic mail to <devs@brlcad\&.org>\&.
.SH "AUTHOR"
.br
.RS 4
Author.
.RE


NEWS
						 -*- coding: utf-8 -*-
BRL-CAD Release Notes
=====================

This document provides a transcript of release notes and user-visible
changes for each release of BRL-CAD.  The individual changes are
listed in chronological order with the most recent changes listed
first and are grouped by release.  Each release also includes optional
descriptive text that emphasizes or further describes significant
changes made.  See document footer for additional details.

----------------------------------------------------------------------
--- 2012-07-XX  Release 7.22.2                                     ---
----------------------------------------------------------------------

* added support for editing 2D sketch objects in archer - Bob Parker
* fixed File->New deprecation error in archer - Sean Morrison
* improved Linux platform application and file icons - Jordi Sayol
* fixed mged interactive object translation keypoint bug - Bob Parker
* fixed corrupted tessellation pipe export bug - Richard Weiss
* output from g-var no longer prints 'fwrite' messages - Sean Morrison
* fixed crash in step-g tool importing AP214 entities - Sean Morrison
* improved 3dm-g output and invalid geometry support - Sean Morrison
* fixed crash revolving empty sketch objects - Sean Morrison
* fixed TIE raytracing bugs shotlining plate-mode BoTs - Sean Morrison
* g_diff gives non-zero return code if .g files differ - Sean Morrison
* fixed typographical errors in numerous tools and docs - Carl Moore
* fixed TIE perpendicular face mesh evaluation failure - Richard Weiss


----------------------------------------------------------------------
--- 2012-06-25  Release 7.22.0                                     ---
----------------------------------------------------------------------

After many months of development, this release includes hundreds of
fixes for source code defects, logic and memory management bugs,
potential security vulnerabilities, and numerous other issues reported
by Coverity Static Analysis.  While having participated in the
Coverity Scan Initiative since inception in 2006, this release marks a
completion milestone with the review and resolution of all reported
issues, more than 2100 overall.  Every defect was inspected, code
modifications were peer-reviewed, critical features were tested, and
user-visible changes were documented.  Along with other on-going
improvement efforts such as treating all compiler-generated warnings
as errors, time and attention continues to be put forth towards
improving BRL-CAD's maintainability and source code quality.

Development of BREP/NURBS representation support within BRL-CAD
continues to mature with this release including ray tracing
improvements, tighter bounding boxes for better performance, progress
converting from traditional implicit+CSG representations to NURBS+CSG,
wireframe improvements, and a thorough verification and validation
(V&V) review for analysis suitability.  The V&V review compared
BRL-CAD's primitives in implicit, polygonal mesh, and NURBS form
evaluating more than a dozen separate tests looking for differences in
rendering, curvature, presented areas, volume, and more.  The review
identified a couple minor issues that were inspected and fixed but
successfully demonstrated the viability of using our BREP/NURBS
implementation for geometric representation and analysis purposes.

This release introduces numerous major feature developments including
the initial implementation of a new physics simulation capability, a
new ambient occlusion rendering mode, and the integration of rtwizard
into Archer.  The simulation system was developed by Abhijit Nandy
under the European Space Agency's 2011 Summer of Code in Space (SOCIS)
program; and it allows BRL-CAD geometry objects to dynamically
interact with each other (e.g., collision detection, friction) or with
the environment (e.g., gravity).  Ambient occlusion is a shading
method that helps add considerable realism and feature definition to
renderings, enabled within rt using the "set ambSamples=128" command.
The rtwizard rendering interface was extended to work as a
command-line tool (useful for scripting) and was integrated within
Archer as the underlying rendering interface.

Conversion support continues to be a development priority with
significant importer enhancements for the STEP, OBJ, COMGEOM, FASTGEN,
and RAW file formats.  The STEP importer in particular receives
several updates, support for importing assembly hierarchies, and is in
the process of being fully ported to Windows.  This release also fixes
more than 40 user-visible bugs including a significant database
corruption issue.  A database corruption issue was discovered and
fixed affecting 64-bit big endian platforms (e.g., IBM Power Series
supercomputers) when writing out ebm, dsp, vol, and hf primitives.

BRL-CAD's documentation continues to be extended, improved, and
translated due to the sustained efforts of several contributors.
Manual page documentation is now available on Windows through the
'brlman' command and within MGED.  Several utilities and commands that
previously had no documentation now have manual pages; many others
received corrections and improvements.  The introductory article
"About BRL-CAD" is now available in Russian, Italian, and Armenian.

* numerous manual page corrections and improvements - Tom Browder
* integration of rtwizard traits into archer - Bob Parker, Cliff Yapp
* new asc2dsp prep tool for displacement map data files - Tom Browder
* fixed a problem with rt writing pix files on Windows - Bob Parker
* added new raw-g raw triangle mesh to BoT importer - Daniel Roßberg
* added Windows man page support (using html) to brlman - Cliff Yapp
* added support for section numbers to brlman script - Cliff Yapp
* new manual pages for asc2dsp, dsp, asc2pix, and asc2g - Tom Browder
* fixed bug creating binary uniform (binunif) objects - Sean Morrison

bench/pixcmp.c
/*                        P I X C M P . C
 * BRL-CAD
 *
 * Copyright (c) 2004-2012 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 *
 */
/** @file pixcmp.c
 *
 * Compute and summarily report differences between two .pix files.
 *
 * This routine operates on a pixel-by-pixel basis, and thus
 * is independent of the resolution of the image.
 *
 */

#include "common.h"

#include <stdlib.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include "bio.h"

#include "bu.h"


/* exit codes for argument processing errors */
#define OPTS_ERROR 127
#define FILE_ERROR 126

/* exit codes for normal operation */
#define OFF_BY_NONE 0
#define OFF_BY_ONE  1
#define OFF_BY_MANY 2


HIDDEN void
usage(const char *name)
{
    const char *unknown = "pixcmp";
    if (!name) {
	name = unknown;
    }
    /* this must be split, ISO C90 has a max static string length of 509 */
    bu_log("Usage: %s [OPTIONS] FILE1 [FILE2 [SKIP1 [SKIP2]]]\n%s", name,
	   "Compare two PIX image files pixel by pixel.\n\n"
	   "  -l   List pixel numbers and values for all pixels that differ.\n"
	   "  -b   Output and process by bytes instead of pixels.\n"
	   "  -i SKIP\n"
	   "       Skip the first SKIP pixels of input (for FILE1 and FILE2)\n");
    bu_log("  -i SKIP1:SKIP2\n"
	   "       Skip the first SKIP1 pixels in FILE1 and SKIP2 pixels in FILE2.\n"
	   "  -s   Silent output.  Only return an exit status.\n\n"
	   "If FILE is `-' or is missing, then input is read from standard input.\n"
	   "If the `-b' option is used, SKIP values are bytes instead of pixels.\n"
	   "Pixel numbers and bytes are indexed linearly from one.\n");
    return;
}


HIDDEN void
handle_i_opt(const char *arg, long *skip1, long *skip2)
{
    const char *endptr = arg;
    if ((arg == NULL) || ((skip1 == NULL) && (skip2 == NULL))) {
	/* nothing to do */
	return;
    }
    while ((endptr != NULL) && (endptr[0] != ':') && (endptr[0] != '\0')) {
	endptr++;
    }
    if (endptr == arg) {
	/* probably empty string */
	if (skip2) {
	    endptr++;
	    *skip2 = strtol(endptr, NULL, 10);
	}
    } else if ((endptr == NULL) || (endptr[0] == '\0')) {
	/* no : found */
	if (skip2) {
	    *skip2 = strtol(arg, NULL, 10);
	}
	if (skip1 && skip2) {
	    *skip1 = *skip2;
	}
    } else if (endptr[0] == ':') {
	/* found : */
	if (skip1) {
	    *skip1 = strtol(arg, NULL, 10);
	}
	if (skip2) {
	    endptr++; /* skip over : */
	    *skip2 = strtol(endptr, NULL, 10);
	}
    } else {
	bu_exit(OPTS_ERROR, "Unexpected input processing [%s]\n", arg);
    }
}


int
main(int argc, char *argv[])
{
    FILE *f1 = NULL;
    FILE *f2 = NULL;

    long matching = 0;
    long off1 = 0;
    long offmany = 0;

    int c;
    int list_pixel_values = 0;
    int print_bytes = 0;
    int silent = 0;
    long f1_skip = 0;
    long f2_skip = 0;

    long int bytes = 0;

    /* process opts */
    while ((c = bu_getopt(argc, argv, "lbi:s")) != -1) {
	switch (c) {
	    case 'l':
		list_pixel_values = 1;
		break;
	    case 'b':
		print_bytes = 1;
		break;
	    case 'i':
		handle_i_opt(bu_optarg, &f1_skip, &f2_skip);
		break;
	    case 's':
		silent = 1;
		break;
	    default:
		bu_log("\n");
		usage(argv[0]);
		exit(OPTS_ERROR);
	}
    }
    argc -= bu_optind;
    argv += bu_optind;

    /* validate what is left over */
    if (argc < 1 || argc > 4) {
	bu_log("ERROR: incorrect number of arguments provided\n\n");
	usage(argv[0]);
	exit(OPTS_ERROR);
    }
    if ((argc > 0 && !argv[0]) || (argc > 1 && !argv[1])) {
	bu_log("ERROR: bad filename\n\n");
	usage(argv[0]);
	exit(OPTS_ERROR);
    }
    if ((argc > 2 && !argv[2]) || (argc > 3 && !argv[3])) {
	bu_log("ERROR: bad skip value\n\n");
	usage(argv[0]);
	exit(OPTS_ERROR);
    }

    /* handle optional SKIP1 and SKIP2 following filenames */
    if (argc > 2) {
	char range[64] = {0};
	if (argc > 3) {
	    snprintf(range, 64, "%s:%s", argv[2], argv[3]);
	} else {
	    snprintf(range, 64, "%s", argv[2]);
	}
	handle_i_opt(range, &f1_skip, &f2_skip);
    }

    /* printf("Skip from FILE1: %ld and from FILE2: %ld\n", f1_skip, f2_skip); */

    if (BU_STR_EQUAL(argv[0], "-")) {
	f1 = stdin;
    } else if ((f1 = fopen(argv[0], "rb")) == NULL) {
	perror(argv[0]);
	exit(FILE_ERROR);
    }
    if ((argc < 2) || BU_STR_EQUAL(argv[1], "-")) {
	f2 = stdin;
    } else if ((f2 = fopen(argv[1], "rb")) == NULL) {
	perror(argv[1]);
	exit(FILE_ERROR);
    }

    if (!print_bytes) {
	f1_skip *= 3;
	f2_skip *= 3;
    }

    /* skip requested pixels/bytes in FILE1 */
    if (f1_skip && fseek(f1, f1_skip, SEEK_SET)) {
	bu_log("ERROR: Unable to seek %ld %s%s in FILE1\n",
	       f1_skip,
	       print_bytes?"byte":"pixel",
	       f1_skip==1?"":"s");
	perror("FILE1 fseek failure");
	exit(FILE_ERROR);
    }

    /* skip requested pixels in FILE2 */
    if (f2_skip && fseek(f2, f2_skip, SEEK_SET)) {
	bu_log("ERROR: Unable to seek %ld %s%s in FILE2\n",
	       f1_skip,
	       print_bytes?"byte":"pixel",
	       f1_skip==1?"":"s");
	perror("FILE2 fseek failure");
	exit(FILE_ERROR);
    }

    /* iterate over the pixels/bytes in the files */
    while ((!feof(f1) && !feof(f2)) &&
	   (!ferror(f1) && !ferror(f2))) {
	register int r1, r2, g1, g2, b1, b2;
	r1 = r2 = g1 = g2 = b1 = b2 = -1;

	r1 = fgetc(f1);
	r2 = fgetc(f2);
	if (feof(f1) || feof(f2)) break;
	bytes++;
	if (!print_bytes) {
	    g1 = fgetc(f1);
	    g2 = fgetc(f2);
	    if (feof(f1) || feof(f2)) break;
	    bytes++;
	    b1 = fgetc(f1);
	    b2 = fgetc(f2);
	    if (feof(f1) || feof(f2)) break;
	    bytes++;
	}

	if ((r1 == r2) && (g1 == g2) && (b1 == b2)) {
	    matching++;
	    continue;
	}

	/* tabulate differing pixels */
	if (((r1 != r2) && (g1 == g2) && (b1 == b2)) ||
	    ((r1 == r2) && (g1 != g2) && (b1 == b2)) ||
	    ((r1 == r2) && (g1 == g2) && (b1 != b2))) {
	    /* off by one channel */
	    if (r1 != r2) {
		if ((r1 > r2 ? r1 - r2 : r2 - r1) > 1) {
		    offmany++;
		} else {
		    off1++;
		}
	    } else if (g1 != g2) {
		if ((g1 > g2 ? g1 - g2 : g2 - g1) > 1) {
		    offmany++;
		} else {
		    off1++;
		}
	    } else if (b1 != b2) {
		if ((b1 > b2 ? b1 - b2 : b2 - b1) > 1) {
		    offmany++;
		} else {
		    off1++;
		}
	    }
	} else {
	    /* off by many */
	    offmany++;
	}

	/* they're different, so print something */
	if (list_pixel_values) {
	    if (print_bytes) {
		printf("%ld %3d %3d\n", bytes, r1, r2);
	    } else {
		printf("%ld\t(%3d, %3d, %3d)\t(%3d, %3d, %3d)\n", bytes / 3, r1, g1, b1, r2, g2, b2);
	    }
	}
    }

    /* print summary */
    if (!silent) {
	printf("pixcmp %s: %8ld matching, %8ld off by 1, %8ld off by many\n",
	       print_bytes?"bytes":"pixels",
	       matching, off1, offmany);
    }

    /* check for errors */
    if (ferror(f1) || ferror(f2)) {
	perror("pixcmp file error");
	return FILE_ERROR;
    }

    /* if files were of different lengths, consider it an error */
    if (feof(f1) != feof(f2)) {
	return FILE_ERROR;
    }

    /* indicate how many differences there were overall */
    if (offmany) {
	return OFF_BY_MANY;
    }
    if (off1) {
	return OFF_BY_ONE;
    }

    /* Success! */
    return OFF_BY_NONE;
}


/*
 * Local Variables:
 * mode: C
 * tab-width: 8
 * indent-tabs-mode: t
 * c-file-style: "stroustrup"
 * End:
 * ex: shiftwidth=4 tabstop=8
 */
